src/Features/Core/Portable/CodeFixesAndRefactorings/FixAllCodeAction.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+using FixAllScope = Microsoft.CodeAnalysis.CodeFixes.FixAllScope;
+
+namespace Microsoft.CodeAnalysis.CodeFixesAndRefactorings
+{
+    internal abstract class FixAllCodeAction : CodeAction
+    {
+        internal readonly IFixAllState FixAllState;
+        private bool _showPreviewChangesDialog;
+
+        protected FixAllCodeAction(
+            IFixAllState fixAllState, bool showPreviewChangesDialog)
+        {
+            FixAllState = fixAllState;
+            _showPreviewChangesDialog = showPreviewChangesDialog;
+        }
+
+        protected abstract bool IsInternalProvider(IFixAllState fixAllState);
+        protected abstract IFixAllContext CreateFixAllContext(IFixAllState fixAllState, IProgressTracker progressTracker, CancellationToken cancellationToken);
+
+        public override string Title
+            => this.FixAllState.Scope switch
+            {
+                FixAllScope.Document => FeaturesResources.Document,
+                FixAllScope.Project => FeaturesResources.Project,
+                FixAllScope.Solution => FeaturesResources.Solution,
+                FixAllScope.ContainingMember => FeaturesResources.Containing_Member,
+                FixAllScope.ContainingType => FeaturesResources.Containing_Type,
+                _ => throw ExceptionUtilities.UnexpectedValue(this.FixAllState.Scope),
+            };
+
+        internal override string Message => FeaturesResources.Computing_fix_all_occurrences_code_fix;
+
+        protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+            => await ComputeOperationsAsync(new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+
+        internal override Task<ImmutableArray<CodeActionOperation>> ComputeOperationsAsync(
+            IProgressTracker progressTracker, CancellationToken cancellationToken)
```

Comment:
nit: can we seal this? (that actually applies to 100% of any overrides you have in this PR).  I find the general inheritance pattern of code-actions pretty hard to reaon about.  every time you can seal an override that cuts off an entire part of my thihnking that has to worry about what it means if this meethod is itself overridden.  thanks!

Reply:
Sounds good, done!